### DISCUSSION TITLE
Can't get filtering working with dynamic schemas

### DISCUSSION THREAD

**Discussion by SeanTAllen** [ChilliCream/hotchocolate#6975] [Q&A]
We have been trying to get this to work for several days. No luck so far. We are starting to think perhaps there is a bug.

We are creating types using an `ITypeModule` that adds types and add fields to query those types to Query root. Everything with this works fine and we can query fine. When trying to add filtering, we get an error.

It very much appears to be the same issue as detailed in [this StackOverflow question](https://stackoverflow.com/questions/71611131/add-filtering-to-schemas-created-dynamically-with-hotchocolate-graphql). However, I tried following the "I seemed to make it work" answer and was unable to do so.

It also appears that it might be the same issue as found in [this discussion](https://github.com/ChilliCream/graphql-platform/discussions/6368). I didn't want to add on to that discussion in case it merely looks the same from outward symptoms.

We are using HotChocolate from F#. Here is extracted code. I am putting together a self contained minimal reproduction that hopefully doesn't end up being "too minimal".

```fsharp
    interface ITypeModule with
      member _.CreateTypesAsync(context, _) =
        task {
          let types = ResizeArray<ITypeSystemMember>()
          let typeDefinition = ObjectTypeDefinition("Query")

          // this gets us our items to be created
          let mdb = database.GetMetadataDatabase()
          let classes = registeredClasses mdb

          for c in classes do
            // this creates the type
            let t = createType c context

            // this is to create the field that is attached to the query root
            let n = graphqlFieldName c.ID

            let fd =
              ObjectFieldDefinition(
                $"%s{n}",
                "",
                TypeReference.Parse($"[%s{n}]"),
                PureResolver = (fun context -> "resolver code here")

            fd.ToDescriptor(context).UseFiltering<MyFilterConvention>((fun f -> f.Name($"%s{n}OperationFilterInput") |> ignore)).ToDefinition()n|> ignore

            typeDefinition.Fields.Add(fd)
            types.Add(t)

          // add the new type to the query root
          types.Add(ObjectTypeExtension.CreateUnsafe(typeDefinition))
          return types :> System.Collections.Generic.IReadOnlyCollection<_>
        }
        |> System.Threading.Tasks.ValueTask<System.Collections.Generic.IReadOnlyCollection<_>>
```

Initially I only had `UseFiltering()` but added a filter convention based on the [stackoverflow answer](https://stackoverflow.com/a/72739455).

That code adds:

```fsharp
  type MyFilterConvention() =
    inherit FilterConvention()

    override _.Configure(descriptor) =
      descriptor.AddDefaults() |> ignore

      descriptor.Provider(Expressions.QueryableFilterProvider((fun f -> f.AddDefaultFieldHandlers() |> ignore)))
      |> ignore
```

The setup for our app is:

```fsharp
    let dbProvider = DatabaseProvider(opts.Configfile)

    let builder = WebApplication.CreateBuilder([||])

    builder.Services
      .AddSingleton(dbProvider)
      .AddSingleton<ClassObjectProvider>()
      .AddGraphQLServer()
      .AddConvention<IFilterConvention, MyFilterConvention>()
      .AddTypeModule<ClassObjectProvider>()
      .AddQueryType<Query>()
      .AddMutationType<Mutation>()
      .AddMutationConventions()
      .AddFiltering()
      .AddSorting()
      .AddFSharpTypeConverters()
      .InitializeOnStartup()
    |> ignore

    let app = builder.Build()
    app.MapGraphQL() |> ignore
    app.Run()

    0
```

We've tried a number of approaches to fix and in almost all cases we end up with a stacktrace like:

```
fail: Microsoft.Extensions.Hosting.Internal.Host[11]
      Hosting failed to start
      HotChocolate.SchemaException: For more details look at the `Errors` property.
      
      1. No default filter convention found. Call `AddFiltering()` on the schema builder.
      
         at HotChocolate.Data.FilterDescriptorContextExtensions.<>c__DisplayClass1_0.<GetFilterConvention>b__0()
         at HotChocolate.Types.Descriptors.DescriptorContext.GetConventionOrDefault[T](Func`1 defaultConvention, String scope)
         at HotChocolate.Data.FilterDescriptorContextExtensions.GetFilterConvention(IDescriptorContext context, String scope)
         at HotChocolate.Types.FilterObjectFieldDescriptorExtensions.<>c__DisplayClass5_0.<UseFiltering>b__1(IDescriptorContext c, ObjectFieldDefinition definition)
         at HotChocolate.Types.Descriptors.DescriptorBase`1.<>c__DisplayClass19_0.<OnBeforeCreate>b__0(IDescriptorContext c, IDefinition d)
         at HotChocolate.Types.Descriptors.Definitions.CreateConfiguration.Configure(IDescriptorContext context)
         at HotChocolate.Types.Descriptors.DescriptorBase`1.CreateDefinition()
         at HotChocolate.Types.Descriptors.DescriptorExtensions.ToDefinition[T](IDescriptor`1 descriptor)
         at UDP.Admin.CLI.Commands.MetadataGraphQL.Pipe #1 input at line 117@117.MoveNext() in /Users/sallen/code/movableink/udp-playground/src/Admin.CLI/Commands.GraphQL.Metadata.fs:line 165
         at HotChocolate.Execution.RequestExecutorResolver.TypeModuleChangeMonitor.TypeModuleEnumerable.GetAsyncEnumerator(CancellationToken cancellationToken)+MoveNext()
         at HotChocolate.Execution.RequestExecutorResolver.TypeModuleChangeMonitor.TypeModuleEnumerable.GetAsyncEnumerator(CancellationToken cancellationToken)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()
         at HotChocolate.Execution.RequestExecutorResolver.CreateSchemaAsync(ConfigurationContext context, RequestExecutorSetup setup, RequestExecutorOptions executorOptions, IServiceProvider schemaServices, TypeModuleChangeMonitor typeModuleChangeMonitor, CancellationToken cancellationToken)
         at HotChocolate.Execution.RequestExecutorResolver.CreateSchemaAsync(ConfigurationContext context, RequestExecutorSetup setup, RequestExecutorOptions executorOptions, IServiceProvider schemaServices, TypeModuleChangeMonitor typeModuleChangeMonitor, CancellationToken cancellationToken)
         at HotChocolate.Execution.RequestExecutorResolver.CreateSchemaServicesAsync(ConfigurationContext context, RequestExecutorSetup setup, CancellationToken cancellationToken)
         at HotChocolate.Execution.RequestExecutorResolver.GetRequestExecutorNoLockAsync(String schemaName, CancellationToken cancellationToken)
         at HotChocolate.Execution.RequestExecutorResolver.GetRequestExecutorAsync(String schemaName, CancellationToken cancellationToken)
         at HotChocolate.AspNetCore.Warmup.ExecutorWarmupService.ExecuteAsync(CancellationToken stoppingToken)
         at Microsoft.Extensions.Hosting.Internal.Host.<StartAsync>b__15_1(IHostedService service, CancellationToken token)
         at Microsoft.Extensions.Hosting.Internal.Host.ForeachService[T](IEnumerable`1 services, CancellationToken token, Boolean concurrent, Boolean abortOnFirstException, List`1 exceptions, Func`3 operation)
Unhandled exception. HotChocolate.SchemaException: For more details look at the `Errors` property.
```

Any assistance would be appreciated. I should be able to have a self-contained reproduction sometime tomorrow.

**Comment by SeanTAllen** [comment]
Here's a minimal reproduction that @pblasucci put together.

[typemodule.zip](https://github.com/ChilliCream/graphql-platform/files/14530122/typemodule.zip)

**Comment by SeanTAllen** [comment]
@jimitndiaye I saw on https://github.com/ChilliCream/graphql-platform/discussions/6272 that you appear to have gotten this working. Could you share any insights here?

**Reply by jimitndiaye** [reply]
I'd suggest adding a call to ```AddProjections``` before calling ```AddFiltering```

**Reply by jimitndiaye** [reply]
Also if the entity/record types are dynamic as well you may need to create a custom class inheriting from ```FilterInputType``` that generates the appropriate filter input fields based on the dynamic record type.

**Reply by SeanTAllen** [reply]
@jimitndiaye did you have existing .net types for the types you were adding to the schema using the dynamic schema API?

**Reply by SeanTAllen** [reply]
I tried adding a called to `AddProjections`. It had no impact and we still get the `No default filter convention found. Call `AddFiltering()` on the schema builder.`

What did you do to get filtering working with a dynamically created schema?

**Reply by jimitndiaye** [reply]
> @jimitndiaye did you have existing .net types for the types you were adding to the schema using the dynamic schema API?

I did not. I had a type that was effectively a dictionary of key-value pairs.

**Reply by jimitndiaye** [reply]
> I tried adding a called to `AddProjections`. It had no impact and we still get the `No default filter convention found. Call `AddFiltering()` on the schema builder.`
> 
> What did you do to get filtering working with a dynamically created schema?

AddFiltering registers the built-in queryable filter convention as default. My implementation did not use that. Instead I had a custom filter provider. The implementation took alot of inspiration from Hotchocolate's mongodb query provider. Thus I provided custom FilterInputType implementations as well as similar implementations for projections and sorting. You may not need all that since your datasource seems to be based on Queryable (EF?).

You will definitely need custom FilterInputType implementations though since you are basing your filters on dynamic types - the default implementation uses reflection to discover the filter fields.

The error message is suggesting it can't find the default filter convention though which should have been registered by AddFiltering. You may try using the overload that takes the name of a specific  provider.

**Reply by SeanTAllen** [reply]
Our data source is not based on Queryable. So the approach you took might be what we require.

We were attempting to do it this way because we have a few types that backed by .net types and we are using an annotation-first approach.

So, with your approach, when you were creating types, did you have something that was equivalent to:

`fd.ToDescriptor(context).UseFiltering<MyFilterConvention>((fun f -> f.Name($"%s{n}OperationFilterInput") |> ignore)).ToDefinition()n|> ignore`

To set up the filtering? I've been wondering if the `ToDefinition` at the end is correct as it is what causes the error. Without it, we simply don't get any filtering. So, perhaps the `ToDefinition` call shouldn't be done and we otherwise setting up filtering wrong and our error is a red herring.

One reading of your answer suggests that could be the case.

**Comment by jimitndiaye** [comment]
Our startup is as follows:
```
            services.AddGraphQLServer()
                .ModifyOptions(o => o.EnableOneOf = true)
                .AddAuthorization()
                .AddQueryType<Query>()
                .AddMutationType<Mutation>()
                .AddConvention<INamingConventions, NamingConventions>()
                .AddTypeModule<ConnectorTypeModule>()
                .AddHttpRequestInterceptor<RequestInterceptor>()
                .AddInMemorySubscriptions()
                .AddDataverseProjections()
                .AddDataverseFiltering()
                .AddDataverseSorting()
                .AddDataversePagingProviders()
                .AddProjections()
                .AddFiltering()
                .AddSorting()
                .AddTypeConverter<DateTimeOffset, DateTime>(t => t.UtcDateTime)
                .AddTypeConverter<DateTime, DateTimeOffset>(t => t.Kind == DateTimeKind.Unspecified
                    ? DateTime.SpecifyKind(t, DateTimeKind.Unspecified)
                    : t)
                .InitializeOnStartup();

```
The AddDataverse*** functions are custom extension methods that do same as the equivalent ```AddFiltering``` but for our datasource. Like I said the mongodb implementation in the repository is a good place to start - you can then tweak it to target your own datasource.

In the type module we have sections of code like:
```
var queryExtensionDescriptor = new ObjectTypeDefinition(OperationTypeNames.Query)
                .ToDescriptor(context);
            queryExtensionDescriptor
                .Field("connectors")
                .Description("The list of connectors currently provided and/or enabled")
                .Type(new NonNullType(connectorListType))
                .Resolve(async ctx =>
                {
                    await SetUserContextDataAsync(ctx);
                    return new();
                });
            var queryExtensionDefinition = queryExtensionDescriptor.ToDefinition();
            types.Add(ObjectTypeExtension.CreateUnsafe(queryExtensionDefinition));
```
That ```types``` variable is a LIst<ITypeSystemMember> and is eventually returned from ```CreateTypesAsync```.

For the actual query fields we register them as such: 
```
descriptor.Field(entity.Metadata.LogicalCollectionName?.Trim() ?? $"{entity.Metadata.LogicalName}_search") // e.g. "contacts"
                .Type($"[{entity.TypeName()}!]")
                .Description($"Allows retrieval of a list of {entity.Metadata.LogicalName} records based on the provided filtering, sorting and paging parameters")
                //.UsePaging(
                //    options: new() { MaxPageSize = 200, DefaultPageSize = 30, AllowBackwardPagination = true, ProviderName = Shared.Constants.Connectors.Dataverse.ConnectorName },
                //    connectionName: entity.TypeName() + "_"
                //)
                .UseProjection(Shared.Constants.Connectors.Dataverse.ConnectorName)
                .UseDataverseFiltering(entity.FilterInputType)
                //.UseDataverseSorting(entity.SortInputType)
                .Resolve((resolverContext, _) =>
                {
                    // details omitted. Here's where you would talk to your datasource and return an IExecutable<T> (see mongodb implementation)
                });
```
You'll notice the use of custom FilterInputType for each entity type. Basically we inherit from ```FilterInputType``` and based on our metadata add filter input fields of the appropriate type:
```
public class EntityFilterInputType : FilterInputType
{
    private readonly Func<string, EntityType?> _entityLookup;
    private readonly EntityType _entityType;
    private readonly string _typeName;

    public EntityFilterInputType(EntityType entityType, Func<string, EntityType?> entityLookup)
    {
        _entityType = entityType;
        _entityLookup = entityLookup;
        _typeName = _entityType.TypeName() + "_Filter_Input";
        Name = _typeName;
        FilterableAttributes = _entityType.Attributes
            .Where(a => a.IsValidForAdvancedFind?.Value == true)
            .OrderBy(x => x.LogicalName)
            .ToArray();
    }
    internal AttributeMetadata[] FilterableAttributes { get; }
    protected override void Configure(IFilterInputTypeDescriptor descriptor)
    {
        base.Configure(descriptor);
        try
        {
            //descriptor.BindFieldsExplicitly();
            descriptor.Name(_typeName);
            foreach (var attribute in FilterableAttributes)
            {
                var isNullable = attribute.RequiredLevel.Value is not (AttributeRequiredLevel.SystemRequired
                    or AttributeRequiredLevel.ApplicationRequired);
                var attributeLogicalName = attribute.LogicalName;
                var field = attribute switch
                {
                    BigIntAttributeMetadata or {AttributeType: AttributeTypeCode.BigInt} => isNullable
                        ? descriptor
                            .Field(attributeLogicalName)
                            .Type<ComparisonOperationFilterInputType<long?>>()
                        : descriptor
                            .Field(attributeLogicalName)
                            .Type<ComparisonOperationFilterInputType<long>>(),
                    BooleanAttributeMetadata or {AttributeType: AttributeTypeCode.Boolean} => descriptor
                        .Field(attributeLogicalName)
                        .Type<BooleanOperationFilterInputType>(),

                    DateTimeAttributeMetadata {Format: DateTimeFormat.DateOnly} => isNullable
                        ? descriptor
                            .Field(attributeLogicalName)
                            .Type<ComparisonOperationFilterInputType<DateOnly?>>()
                        : descriptor
                            .Field(attributeLogicalName)
                            .Type<ComparisonOperationFilterInputType<DateOnly>>(),
                    DateTimeAttributeMetadata or {AttributeType: AttributeTypeCode.DateTime} => isNullable
                        ? descriptor
                            .Field(attributeLogicalName)
                            .Type<ComparisonOperationFilterInputType<DateTime?>>()
                        : descriptor
                            .Field(attributeLogicalName)
                            .Type<ComparisonOperationFilterInputType<DateTime>>(),
                    DecimalAttributeMetadata or {AttributeType: AttributeTypeCode.Decimal} => isNullable
                        ? descriptor
                            .Field(attributeLogicalName)
                            .Type<ComparisonOperationFilterInputType<decimal?>>()
                        : descriptor
                            .Field(attributeLogicalName)
                            .Type<ComparisonOperationFilterInputType<decimal>>(),
                    DoubleAttributeMetadata or {AttributeType: AttributeTypeCode.Double} => isNullable
                        ? descriptor
                            .Field(attributeLogicalName)
                            .Type<ComparisonOperationFilterInputType<double?>>()
                        : descriptor
                            .Field(attributeLogicalName)
                            .Type<ComparisonOperationFilterInputType<double>>(),
                    IntegerAttributeMetadata or {AttributeType: AttributeTypeCode.Integer} => isNullable
                        ? descriptor
                            .Field(attributeLogicalName)
                            .Type<ComparisonOperationFilterInputType<int?>>()
                        : descriptor
                            .Field(attributeLogicalName)
                            .Type<ComparisonOperationFilterInputType<int>>(),
                    LookupAttributeMetadata or
                    {
                        AttributeType: AttributeTypeCode.Lookup or AttributeTypeCode.Owner or AttributeTypeCode.Customer
                    } => isNullable
                        ? descriptor
                            .Field(attributeLogicalName)
                            .Type<ComparisonOperationFilterInputType<Guid?>>()
                        : descriptor
                            .Field(attributeLogicalName)
                            .Type<ComparisonOperationFilterInputType<Guid>>(),
                    MemoAttributeMetadata or StringAttributeMetadata or
                        {AttributeType: AttributeTypeCode.Memo or AttributeTypeCode.String} =>
                        isNullable
                            ? descriptor
                                .Field(attributeLogicalName)
                                .Type<NullableStringOperationFilterInputType>()
                            : descriptor
                                .Field(attributeLogicalName)
                                .Type<StringOperationFilterInputType>(),
                    //.Type(typeof(string)),
                    MoneyAttributeMetadata or {AttributeType: AttributeTypeCode.Money} =>
                        isNullable
                            ? descriptor
                                .Field(attributeLogicalName)
                                .Type<ComparisonOperationFilterInputType<decimal?>>()
                            : descriptor
                                .Field(attributeLogicalName)
                                .Type<ComparisonOperationFilterInputType<decimal>>(),
                    UniqueIdentifierAttributeMetadata or
                        {AttributeType: AttributeTypeCode.Uniqueidentifier} => isNullable
                            ? descriptor
                                .Field(attributeLogicalName)
                                .Type<ComparisonOperationFilterInputType<Guid?>>()
                            : descriptor
                                .Field(attributeLogicalName)
                                .Type<ComparisonOperationFilterInputType<Guid>>(),
                    // details omitted ...
                    _ => null
                };
               // other implementation details
                }
                else
                    field?.Description($"Filter by '{attributeLogicalName}'");

            }
        }
        catch (Exception ex)
        {
            ex.Trace()
            throw;
        }
    }
}
```
Ultimately, I recommend you look at the MongoDb code within Hotchocolate and see about replacing the mongodb-specific implementations with custom code that talks to your own datasource.

**Reply by SeanTAllen** [reply]
Thanks Jim. I'm going to spend some time absorbing what you've posted. I will most likely adapt it to what we are doing and report back. 

Thank you very much for sharing your knowledge.

**Reply by SeanTAllen** [reply]
@jimitndiaye 

This line:

`.UseDataverseFiltering(entity.FilterInputType)`

The `entity.FilterInputType` I assume that is always an instance of EntityFilterInputType? Are the instances the same across all classes you register or do they vary based on the class being created?

Is this detail important to getting things working or merely a detail of your domain?

**Reply by SeanTAllen** [reply]
@jimitndiaye are you referring to this MongoDB code? https://github.com/ChilliCream/graphql-platform/tree/2c90ea18962c6ccdb21fed99b1f6da28d0a53b54/src/HotChocolate/MongoDb

There's a lot of in there, did you use most of that for yours or is there a subset that you used as a basis for your code?

**Reply by SeanTAllen** [reply]
Also, what does `UseDataverseFiltering` actually do? I don't see an obvious equivalent in the MongoDB code.

**Reply by jimitndiaye** [reply]
> @jimitndiaye 
> 
> This line:
> 
> `.UseDataverseFiltering(entity.FilterInputType)`
> 
> The `entity.FilterInputType` I assume that is always an instance of EntityFilterInputType? Are the instances the same across all classes you register or do they vary based on the class being created?
> 
> Is this detail important to getting things working or merely a detail of your domain?

Yes it is an instance of ```EntityFilterInputType```. Each Entity in our dynamic schema has an associated instance of the filter input type that models the filter fields for that entity. You may have noticed that ```EntityFilterInputType``` takes an instance of ```EntityType``` which it uses to determine which fields from the original entity to create filters for. This is just an implementation detail though. 

The call to ```UseDataverseFiltering``` is very important - it is equivalent to the ```UseMongoFiltering``` extension method except modified to take an instance of a ```FilterInputType``` instead of a type as well as default to my custom provider. It is basically the same as calling ```UseFiltering``` and passing in the name of a custom provider. Dataverse just happens to be the name of my provider hence the name of the function.

**Reply by jimitndiaye** [reply]
> @jimitndiaye are you referring to this MongoDB code? https://github.com/ChilliCream/graphql-platform/tree/2c90ea18962c6ccdb21fed99b1f6da28d0a53b54/src/HotChocolate/MongoDb
> 
> There's a lot of in there, did you use most of that for yours or is there a subset that you used as a basis for your code?

Yes.
I need most of it, except for the Mongodb-specific code which I replaced with code targeting my data source.

Specifically I needed code for filtering, projections, sorting and paging. You may not need all of those features. At a minimum I'd expect you need filtering and projections.

**Reply by SeanTAllen** [reply]
@jimitndiaye Do you need that code for filtering or for supporting other things you are doing like projections? I'm assuming that what was needed to get filtering working was a much smaller subset?

**Reply by jimitndiaye** [reply]
Filtering needs projections. Projections are what allow you to determine which fields have been selected from the selection set and thus the filtering middleware needs the projections middleware.

**Reply by michaelstaib** [reply]
The filter middleware has no dependency on projections. You can use filtering without projections. When using MongoDB I recommend not using projections at all as performance is worse if not all fields are indexed.

**Reply by SeanTAllen** [reply]
@jimitndiaye what does `.AddDataverseFiltering()` do when you call it? I'm trying to understand how what you did fixed the issue and that's a key part that I don't understand.

**Reply by SeanTAllen** [reply]
@jimitndiaye also, I can't find `UseMongoFiltering` anywhere. Do you have a pointer to what you meant by that? Was it in an order version of the code perhaps? I also searched for `UseMongoDbFiltering`.

**Reply by jimitndiaye** [reply]
It is basically the same as ```UseFiltering``` but modified to take an instance of ```FilterInputType```:
```C#
namespace HotChocolate.Types;

public static class DataverseFilterObjectFieldDescriptorExtensions
{
    private static readonly MethodInfo _factoryTemplate =
        typeof(DataverseFilterObjectFieldDescriptorExtensions)
            .GetMethod(nameof(CreateMiddleware), BindingFlags.Static | BindingFlags.NonPublic)!;

    /// <summary>
    ///     Registers the middleware and adds the arguments for Dataverse Filtering
    /// </summary>
    /// <param name="descriptor">
    ///     The field descriptor where the arguments and middleware are
    ///     applied to
    /// </param>
    /// <param name="filterInputType">The <see cref="EntityFilterInputType" /> instance</param>
    public static IObjectFieldDescriptor UseDataverseFiltering(
        this IObjectFieldDescriptor descriptor,
        EntityFilterInputType filterInputType)
    {
        if (descriptor is null) throw new ArgumentNullException(nameof(descriptor));
        if (filterInputType is null) throw new ArgumentNullException(nameof(filterInputType));
        if (filterInputType is {FilterableAttributes.Length: > 0})
            return UseFilteringInternal(descriptor, filterInputType,
                Constants.Connectors.Dataverse.ConnectorName);
        return descriptor;
    }

    private static IObjectFieldDescriptor UseFilteringInternal(
        IObjectFieldDescriptor descriptor,
        IFilterInputType filterTypeInstance,
        string? scope)
    {
        FieldMiddlewareDefinition placeholder =
            new(_ => _ => default, key: WellKnownMiddleware.Filtering);

        var argumentPlaceholder =
            "_" + Guid.NewGuid().ToString("N", CultureInfo.InvariantCulture);

        descriptor.Extend().Definition.MiddlewareDefinitions.Add(placeholder);

        descriptor
            .Extend()
            .OnBeforeCreate(
                (c, definition) =>
                {
                    var argumentTypeReference = TypeReference.Create(filterTypeInstance, scope);
                    var argumentDefinition = new ArgumentDefinition
                    {
                        Name = argumentPlaceholder,
                        Type = argumentTypeReference
                    };
                    definition.Arguments.Add(argumentDefinition);

                    definition.Configurations.Add(
                        new CompleteConfiguration<ObjectFieldDefinition>(
                            (context, def) =>
                                CompileMiddleware(
                                    context,
                                    def,
                                    argumentTypeReference,
                                    placeholder,
                                    scope),
                            definition,
                            ApplyConfigurationOn.BeforeCompletion,
                            argumentTypeReference,
                            TypeDependencyFulfilled.Completed));

                    argumentDefinition.Configurations.Add(
                        new CompleteConfiguration<ArgumentDefinition>(
                            (context, argDef) =>
                                argDef.Name =
                                    context.GetFilterConvention(scope).GetArgumentName(),
                            argumentDefinition,
                            ApplyConfigurationOn.BeforeNaming));
                });

        return descriptor;
    }

    private static void CompileMiddleware(
        ITypeCompletionContext context,
        ObjectFieldDefinition definition,
        TypeReference argumentTypeReference,
        FieldMiddlewareDefinition placeholder,
        string? scope)
    {
        var type = context.GetType<IFilterInputType>(argumentTypeReference);
        var convention = context.DescriptorContext.GetFilterConvention(scope);

        var fieldDescriptor = ObjectFieldDescriptor.From(context.DescriptorContext, definition);
        convention.ConfigureField(fieldDescriptor);

        var factory = _factoryTemplate.MakeGenericMethod(type.EntityType.Source);
        var middleware = (FieldMiddleware) factory.Invoke(null,
            new object[]
            {
                convention
            })!;
        var index = definition.MiddlewareDefinitions.IndexOf(placeholder);
        definition.MiddlewareDefinitions[index] =
            new(middleware, key: WellKnownMiddleware.Filtering);
    }

    private static FieldMiddleware CreateMiddleware<TEntity>(
        IFilterConvention convention) =>
        convention.CreateExecutor<TEntity>();
}
```
Generally simply calling the builtin ```UseFiltering``` function would be fine but I needed to be able to pass in the specifc instance of ```FilterInputType``` which the built-in function does not support.

**Reply by gerneio** [reply]
Note for others: instead of re-implementing the internals of the private static method `FilterObjectFieldDescriptorExtensions.UseFiltering` (as shown above by @jimitndiaye), if you're ok with using reflection, the below also works quite well, and much less repeat code:

```cs
    public static IObjectFieldDescriptor UseCustomFiltering(this IObjectFieldDescriptor descriptor, IFilterInputType filterInputType, string? scope = null)
    {
        ArgumentNullException.ThrowIfNull(descriptor);
        ArgumentNullException.ThrowIfNull(filterInputType);

        // private static IObjectFieldDescriptor UseFiltering(IObjectFieldDescriptor descriptor, Type? filterType, ITypeSystemMember? filterTypeInstance, string? scope)
        var _UseFilteringInternal = typeof(FilterObjectFieldDescriptorExtensions)
            .GetMethod("UseFiltering", BindingFlags.Static | BindingFlags.NonPublic);

        ArgumentNullException.ThrowIfNull(_UseFilteringInternal);

        _UseFilteringInternal.Invoke(null, [descriptor, filterInputType.GetType(), filterInputType, scope]);

        return descriptor;
    }
```

Also easy to add similar extensions for `UseSorting` overrides
```cs
    public static IObjectFieldDescriptor UseCustomSorting(this IObjectFieldDescriptor descriptor, ISortInputType sortInputType, string? scope = null)
    {
        ArgumentNullException.ThrowIfNull(descriptor);
        ArgumentNullException.ThrowIfNull(sortInputType);

        // private static IObjectFieldDescriptor UseSortingInternal(IObjectFieldDescriptor descriptor, Type? sortType, ITypeSystemMember? sortTypeInstance, string? scope)
        var _UseSortingInternal = typeof(SortObjectFieldDescriptorExtensions)
            .GetMethod("UseSortingInternal", BindingFlags.Static | BindingFlags.NonPublic);

        ArgumentNullException.ThrowIfNull(_UseSortingInternal);

        _UseSortingInternal.Invoke(null, [ descriptor, sortInputType.GetType(), sortInputType, scope ]);

        return descriptor;
    }
```

Both tested to work on HC 13. Of course, by going the reflection route, we're probably opening ourselves up to issues when updating to latest HC later down the road if any internals change. Would be nice if we could get these as built-in overloads. Doesn't seem like it would cause any heartaches to add.

Also quick note (mostly for my future self in case I need to come back to this later), depending on your situation, but assuming you're dealing with dynamic schemas/filtering/sorting (like myself and the others in this discussion), it might be suitable to use the `UseFiltering`/`UseSorting` overloads with the configure action param. They require passing in a generic type, which I'd assume would be your data model normally, but if you're dealing with a dynamic schema you probably won't have one, so something like this also seemed to work well:

```cs
    .UseFiltering<object>(filterDescriptor =>
    {
        filterDescriptor.BindFieldsExplicitly(); // Ignores automatic building of fields based on the generic param
        filterDescriptor.Name("{unique_filter_name}");

        foreach (var field in fields)
        {
            filterDescriptor.Field(field.Name)
                .Type<StringOperationFilterInputType>()
                .Extend()
                .OnBeforeCreate(e =>
                {
                    e.Handler = new CustomFilterFieldHandler();
                });
        }
    })
    .UseSorting<object>(sortDescriptor =>
    {
        sortDescriptor.BindFieldsExplicitly(); // Ignores automatic building of fields based on the generic param
        sortDescriptor.Name("{unique_sort_name}");

        foreach (var field in fields)
        {
            sortDescriptor.Field(field.Name)
                .Type<DefaultSortEnumType>()
                .Extend()
                .OnBeforeCreate(e =>
                {
                    e.Handler = new CustomSortFieldHandler();
                });
        }
    })
```

Downsides being that if you need to tie into any of the other lifecycle methods available when using a custom type derived from `FilterInputType`/`SortInputType`, then you're out of luck since you're only provided a hook to `Configure`, but works well if you have a simple dynamic case.

Hope this helps someone else cause I spent WAY too much time on this. Thanks @jimitndiaye for all the above details and @SeanTAllen for creating this thread which helped point me into the right direction.

P.S. Last note: if you're adding multiple different types of dynamic Filtering/Sorting under the same dynamic schema, it's crucial that the `descriptor.Name()` method is called with a unique name per filter/sort needed. Might be obvious, but the errors received can sometimes be a bit confusing on the main culprit.

**Reply by SeanTAllen** [reply]
@gerneio were you able to get the `OnBeforeCreate` approach working? I was trying this when we needed to expand what we were doing, but I am stuck at having a handler so it doesn't complain about not having the handler, but I can't actually get the handler to do anything. If you got it working with this approach, could you share an example of a working `CustomFilterFieldHandler`?

**Comment by SeanTAllen** [comment]
If I understand what @jimitndiaye did. And what the MongoDB does, this appears to set up an alternate filtering system separate from the one that comes from `UseFiltering`. 

I have some indications I could make that work. I also have some indications that there's a bug with making `UseFiltering` work with dynamic types. Or at least a bad error message.

When trying to make `UseFiltering` work, I've repeatedly run into error messages like:

`No filter convention found for scope queryable. Register a convention with AddConvention<IFilterConvention, TYourConvention>("queryable") on the schema builder.`

Where I already have:

`AddConvention<IFilterConvention, CatFilterConvention>("queryable")`

At minimum, there's a misleading error message. Or perhaps there's a HC bug. Or I am missing something that is "obvious" in its non-obviousness.

I would be happy to explore further with an HC maintainer should they take notice of this. Otherwise, it seems like creating a full set of custom filters etc from scratch is probably going to be more fruitful.

**Comment by SeanTAllen** [comment]
After additional debugging, we believe this to be a bug and have opened https://github.com/ChilliCream/graphql-platform/issues/6983

**Comment by jimitndiaye** [comment]
> @jimitndiaye what does `.AddDataverseFiltering()` do when you call it? I'm trying to understand how what you did fixed the issue and that's a key part that I don't understand.

It basically calls ```AddFiltering``` with some defaults setup for my dataprovider. You can find an equivalent implementation in the Mongodb code extension method ```AddMongoFiltering``` - mine is a direct adaptation of theirs.

**Comment by SeanTAllen** [comment]
@jimitndiaye Thanks for all your assistance. I was able to figure out from your work what was causing the error we were encountering and make a small fix in Hot Chocolate Data to address the problem. I'll be opening a PR for that in the not so distant future. Understanding what you were doing with your custom middleware was key to figuring out what we needed to do for our use case of "just needing filtering to work with dynamic schema types".

**Reply by jimitndiaye** [reply]
You're welcome. Glad you found a solution.

**Comment by SeanTAllen** [comment]
For folks who see this in the future, we have a PR open to facilitate the approach we took:

https://github.com/ChilliCream/graphql-platform/pull/7019

I will be creating an example "how to" repository and adding that to this discussion thread before closing this.

**Reply by antoniaelek** [reply]
Hi, I know it's been a while now, but since I am facing a similar issue just now, did you ever end up putting together that how-to?

**Reply by SeanTAllen** [reply]
@antoniaelek I haven't. If I pulled together some F# would be enough to get your going? I think that the minimal example at the beginning plus the PR for a fix that I got merged would be working but I haven't tried it in quite a while.
